urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,223 @@
+            result = self._wait_for_connection(False)
+            if result:
+                break
+            self._timeout_check()
+
+        if result is None:
+            if self._error:
+                raise self._error
+            else:
+                raise socket.error(errno.ECONNREFUSED)
+
+        # Restore the old timeout value and remove it from cleanup.
+        result.settimeout(self._timeout)
+        self._sockets.remove(result)
+        self._selector.unregister(result)
+
+        return result
+
+    def _wait_for_connection(self, last_socket=False):
+        """ Does the actual selecting on the sockets waiting
+        for one to be ready to be written to. """
+        # Make sure we don't have any errored sockets.
+        for sock in self._sockets:
+            self._socket_check(sock)
+
+        # If there are no sockets to select on, then don't try.
+        if not self._sockets:
+            return None
+
+        # If it's the last socket then we can block until timeout
+        if last_socket:
+            if self._timeout is None:
+                select_time = None
+            else:
+                select_time = max(0.0, self._timeout - (current_time() - self._start_time))
+        # Otherwise block for 200ms like the RFC suggests.
+        else:
+            select_time = self._get_select_time()
+
+        ready = self._selector.select(select_time)
+
+        for key, _ in ready:
+            sock = key.fileobj
+
+            # Successfully found a proper connected socket.
+            if self._socket_check(sock):
+                return sock
+
+        return None
+
+    def _create_socket(self, family, socktype, proto, sa):
+        """ Tries it's best to create a socket and register
+        the new socket with the selector. If it fails, no
+        problem, the algorithm continues as usual. """
+        sock = None
+        try:
+            sock = socket.socket(family, socktype, proto)
+            if self._timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
+                self._timeout = sock.gettimeout()
+
+            if self._socket_options:
+                for opt in self._socket_options:
+                    sock.setsockopt(*opt)
+
+            if self._source_address:
+                sock.bind(self._source_address)
+
+            # Set the socket to non-blocking so we can select it.
+            sock.settimeout(0.0)
+
+            error_code = sock.connect_ex(sa)
+            if error_code and error_code not in _ASYNC_ERROR_NUMBERS:
+                self._error = socket.error(error_code)
+            else:
+                self._sockets.append(sock)
+                self._selector.register(sock, EVENT_WRITE)
+        except (socket.error, OSError) as e:
+            self._error = e
+            if sock is not None:
+                try:
+                    sock.close()
+                except (OSError, socket.error):
+                    pass
```

Comment:
I had a question about this, the only two errors I'm aware of from closing something like a socket are interrupts and errors when closing an already closed socket.  Is that something we can ignore or should we still wrap closes like this?